# Model Summary 
Results from the dataset will enable us to predict a malignant tumour in the breast or a benign tumour. Since there are two possible outcomes based on the model to be implememented below, this will be a logistic regression problem. We will assign 1 to a malignant tumour, and 0 to a bening tumour.

Importing the necessary libraries

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.datasets import load_breast_cancer

Initialize our dataset to a pandas dataframe

In [11]:
breast_cancer_dataset = load_breast_cancer()


# Begin preprocessing of our data
breast_cancer_df = pd.DataFrame(data=breast_cancer_dataset['data'], columns=breast_cancer_dataset['feature_names'])

breast_cancer_df['target'] = pd.Series(breast_cancer_dataset.target)

# We will reassign the values to work with our dataset columns 
breast_cancer_dataset.target_names = ['benign', 'malignant']

breast_cancer_df['diagnosis'] = breast_cancer_df.target.replace(dict(enumerate(breast_cancer_dataset.target_names)))


Get a view of dataset fields we will be working with

In [ ]:
breast_cancer_df.head()

mean radius  mean texture  ...  target  diagnosis
0        17.99         10.38  ...       0     benign
1        20.57         17.77  ...       0     benign
2        19.69         21.25  ...       0     benign
3        11.42         20.38  ...       0     benign
4        20.29         14.34  ...       0     benign

[5 rows x 32 columns]

Verify that the extracted data does not have a bias zero value for every row with respect to the target column

In [ ]:
breast_cancer_df.tail()

mean radius  mean texture  ...  target  diagnosis
564        21.56         22.39  ...       0     benign
565        20.13         28.25  ...       0     benign
566        16.60         28.08  ...       0     benign
567        20.60         29.33  ...       0     benign
568         7.76         24.54  ...       1  malignant

[5 rows x 32 columns]

The next step would be to identify the dimesion of the dataset

In [ ]:
print(breast_cancer_df.shape)

(569, 32)


The dimesions are m x n are 569 and 30 respectively where m is number of rows and n number of columns of the subjected dataframe / matrix

Next we check for any null values

In [ ]:
breast_cancer_df.isnull().sum()
breast_cancer_df.isna().sum()

mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
target                     0
diagnosis                  0
dtype: int64

Here we aim to produce categorical data for our model to understand better through encoding


In [31]:

from sklearn.preprocessing import LabelEncoder
X = breast_cancer_dataset.data
Y = breast_cancer_df['diagnosis']
labelencoder_Y = LabelEncoder()
Y = labelencoder_Y.fit_transform(Y)
print(Y)

0         benign
1         benign
2         benign
3         benign
4         benign
         ...    
564       benign
565       benign
566       benign
567       benign
568    malignant
Name: diagnosis, Length: 569, dtype: object


We then split the data into training and test data. 25% for testing, 75% for training

In [21]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.25, random_state = 0)

We then scale the data to transform it within certain magnitudes using StandardScaler


In [22]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

We then choose the model for classification. In our case it will be Logistic Regression. We start by making a classifier instance for the training data
 and fitting it with the data

In [33]:
from sklearn.linear_model import LogisticRegression
logReg = LogisticRegression(random_state = 0)
logReg.fit(X_train, Y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=0, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

We then make predictions on the test data and calculate the accuracy score

In [34]:
preds = logReg.predict(X_test)
score = logReg.score(X_test, Y_test)
print(score)

0.958041958041958


We then use the confusion matrix below to evaluate the  model. As said earlier 0 is benign and 1 is malignant

In [38]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(Y_test, preds)
cm
confusion_df = pd.DataFrame(confusion_matrix(Y_test,preds),
             columns=["Predicted Class " + str(breast_cancer_dataset.target_names) for breast_cancer_dataset.target_names in [0,1]],
             index = ["Class " + str(breast_cancer_dataset.target_names) for breast_cancer_dataset.target_names in [0,1]])
print(confusion_df)

         Predicted Class 0  Predicted Class 1
Class 0                 50                  3
Class 1                  3                 87


We can then print the classification report

In [40]:
from sklearn.metrics import classification_report
print(classification_report(Y_test, preds))

              precision    recall  f1-score   support

           0       0.94      0.94      0.94        53
           1       0.97      0.97      0.97        90

    accuracy                           0.96       143
   macro avg       0.96      0.96      0.96       143
weighted avg       0.96      0.96      0.96       143

